File: python/src/bench_sdk/runtime_gateway.py
"""Guest-side file transport for an offline sandbox. No credentials required."""
import json
import os
from pathlib import Path
import time

class BenchGateway:
    def __init__(self):
        self.path=Path(os.environ['BENCH_GATEWAY_PATH']);self.sequence=0
    def model(self,messages,tools=None):
        n=self.sequence;self.sequence+=1
        payload={'operation':'model','messages':messages}
        if tools:payload['tools']=tools
        temporary=self.path/f'request-{n}.tmp'
        temporary.write_text(json.dumps(payload))
        temporary.replace(self.path/f'request-{n}.json')
        deadline=time.monotonic()+115
        response=self.path/f'response-{n}.json'
        while not response.exists():
            if time.monotonic()>deadline:raise TimeoutError('Gateway response unavailable')
            time.sleep(.05)
        value=json.loads(response.read_text())
        if 'error' in value:raise RuntimeError(value['error'])
        return value['message']

File: src/runtime-gateway.ts
import { readFile, writeFile, rename } from 'node:fs/promises';
import path from 'node:path';
/** Guest-side transport for Bench's offline hosted runtime. */
export class RuntimeGateway {
  private sequence = 0;
  constructor(private directory = process.env.BENCH_GATEWAY_PATH) {
    if (!directory) throw new Error('Bench gateway is available only inside a configured runtime');
  }
  async model(messages: Record<string,unknown>[], tools?: Record<string,unknown>[]): Promise<Record<string,unknown>> {
    const id=this.sequence++;
    const prefix=path.join(this.directory!,`request-${id}`);
    await writeFile(`${prefix}.tmp`,JSON.stringify({operation:'model',messages,...(tools?{tools}:{})}));
    await rename(`${prefix}.tmp`,`${prefix}.json`);
    const deadline=Date.now()+115000;
    while (Date.now()<deadline) {
      let raw: string;
      try { raw=await readFile(path.join(this.directory!,`response-${id}.json`),'utf8'); }
      catch(error) {
        if ((error as NodeJS.ErrnoException).code!=='ENOENT') throw error;
        await new Promise(resolve=>setTimeout(resolve,50)); continue;
      }
      const result=JSON.parse(raw);
      if (result.error) throw new Error(result.error);
      return result.message;
    }
    throw new Error('Bench gateway deadline exceeded');
  }
}
